### PR TITLE
[WIP] RWKV4Neo the RNN and GPT Hybrid Model

### PR DESCRIPTION
**Pull request by ArEnSc**
# What does this PR do?
Adds the model from issue
Fixes #  (https://github.com/huggingface/transformers/issues/20737)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@younesbelkada 
@ArthurZucker

### CODE REVIEW

**Comment by younesbelkada** [comment]
Hi @ArEnSc !
Thanks for starting over the PR 💪 
Let us know whenever you need help with @ArthurZucker !

**Comment by ArEnSc** [comment]
> Hi @ArEnSc ! Thanks for starting over the PR 💪 Let us know whenever you need help with @ArthurZucker !

Will do still doing some research, just figured out how the training notebook works, model executes in notebook so that's a positive

**Comment by ArEnSc** [comment]
Update: tracing the model and came up with a state based api for the RNN inference mode on my own code base to experiment with

**Comment by younesbelkada** [comment]
Thanks a lot for the status update! Feel free to ping whenever you need help

**Comment by xloem** [comment]
- The template appears to be from a T5 style model. Here, the encoder hidden states could be the passed state in rwkv. It will take some algebra and tests to add those to its GPT form from its RNN form.
- [ ] The model does not appear to use position and token type embeddings, nor dropout or layernorm after them. It could be simplifying to remove `Rwkv4NeoEmbeddings` and replace with a simple `nn.Embedding`, optionally migrating `if inputs_embeds is None:` into the encoder class, so the user can provide them manually.
- [ ] review or improve more of model file

**Comment by Lundez** [comment]
@ArEnSc do you need any help?

**Comment by younesbelkada** [comment]
Hi @ArEnSc,
Can you share with use your discord handle? Thanks!

**Comment by ArEnSc** [comment]
> Hi @ArEnSc, Can you share with us your discord handle? Thanks!

ARENSC#5905
yeah still working on it haha it will be a while

**Comment by ArEnSc** [comment]
Working on having GPT Encoder to generate the context and RNN mode inference and sharing weights

**Comment by ArEnSc** [comment]
Deleted a bunch of not needed stuff

**Comment by ArEnSc** [comment]
> @ArEnSc Please let us know if you won't have time to finish this PR. The model is heavily requested as you may see from the linked issue, do you want us to take over this PR and finish this?

Sure yes, sorry been busy at the hospital these days! I think it's probably important that you guys take this on =)
